### DISCUSSION TITLE
Schema caching

### DISCUSSION THREAD

**Discussion by csrster** [phax/ph-schematron#181] [Q&A]
I was wondering if there was a general documentation of how caching works?

The concrete case is that we are validating som xml using the following:

```
ISchematronResource schematronResource = SchematronResourceSCH.fromFile(schematronProfilePath);
.
.
SchematronOutputType schematronOutputType = schematronResource.applySchematronValidationToSVRL(
                    new StreamSource(ourXmlFile));
```
 
The code is running inside an application server (specifically an Apache NiFi custom processor) and we would like to be able to dynamically reload the schematron profile without reloading the application (which would require us to restart NiFi, so far as we understand - their documention isn't great).

Would `SchematronResourceSCH.setUseCache(false) ` do what we want - ie guarantee that the profile is reloaded on each use? The name would suggest so, but I can't find any documentation to confirm it.

Thanks in advance!

**Comment by phax** [comment]
Hi @csrster thanks for that question.

Caching comings in different ways, depending on the source format:
* If the source format is SCH, first the SCH is converted to XSLT, than the XSLT is precompiled into a Java `Templates` object which is cached
* If the source format is XSLT, the XSLT is precompiled into a Java `Templates` object which is cached

Indeed, if you don't want the caching of Schematron, please use `setUseCache (false)`.

But bare in mind, that the time to convert SCH to XSLT as well as the precompilation is time consuming. I recommend that you do the SCH to XSLT conversion at build-time e.g. via the Maven plugin as described in https://github.com/phax/ph-schematron/wiki/Maven-plugin

hth
